onnxruntime/core/providers/cuda/tensor/tile_impl.cu

```diff
@@ -7,140 +7,178 @@
+    }
   }
-  output_data[id] = input_data[input_index];
 }
 
 template <typename T>
-void TileImpl(
-    cudaStream_t stream,
-    const size_t shape_rank,
-    const TArray<fast_divmod>& fdm_input_shape,
-    const TArray<int64_t>& input_stride,
-    const T* input_data,
-    const TArray<fast_divmod>& fdm_output_strides,
-    T* output_data,
-    const size_t N) {
-  int blocksPerGrid = (int)(ceil(static_cast<float>(N) / GridDim::maxThreadsPerBlock));
-  if (shape_rank > MAX_DIMS) {
-    _TileKernel<T><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(
-        shape_rank, fdm_input_shape, input_stride, input_data,
-        fdm_output_strides, output_data, (CUDA_LONG)N);
-  } else {
-    _UnRolledTileKernel<T><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(
-        shape_rank, fdm_input_shape, input_stride, input_data,
-        fdm_output_strides, output_data, (CUDA_LONG)N);
-  }
+void TileImpl(cudaStream_t stream, const size_t shape_rank, const TArray<fast_divmod>& fdm_input_shape,
+              const TArray<int64_t>& input_stride, const T* input_data, const TArray<fast_divmod>& fdm_output_strides,
+              T* output_data, const size_t N) {
+  int blocksPerGrid = static_cast<int>(CeilDiv(N, num_threads_per_block * num_elements_per_thread));
+  _UnRolledTileKernel<T><<<blocksPerGrid, num_threads_per_block, 0, stream>>>(shape_rank, fdm_input_shape, input_stride,
+                                                                              input_data, fdm_output_strides,
+                                                                              output_data, static_cast<CUDA_LONG>(N));
 }
 
 template <typename T>
-__global__ void _TileMemcpyKernel(
-    const T* input_data,
-    const size_t num_input_elements,
-    T* output_data,
-    const size_t N) {
+__global__ void _TileMemcpyKernel(const T* input_data, T* output_data, const CUDA_LONG N, const size_t repeats) {
   CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
-  auto input_index = id % num_input_elements;
-  output_data[id] = input_data[input_index];
+  T input_val = input_data[id];
+  for (size_t i = 0; i < repeats; ++i) {
+    output_data[id] = input_val;
+    id += N;
+  }
 }
 
 template <typename T>
-void TileMemcpyImpl(
-    cudaStream_t stream,
-    const T* input_data,
-    const size_t num_input_elements,
-    T* output_data,
-    const size_t num_output_elements) {
-  int blocksPerGrid = (int)(ceil(static_cast<float>(num_output_elements) / GridDim::maxThreadsPerBlock));
-  _TileMemcpyKernel<<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(
-      input_data, num_input_elements, output_data, (CUDA_LONG)num_output_elements);
+void TileMemcpyImpl(cudaStream_t stream, const T* input_data, T* output_data, const size_t num_input_elements,
+                    const size_t repeats) {
+  using Vec4T = aligned_vector<T, 4>;
+  using Vec2T = aligned_vector<T, 2>;
+  constexpr int vec4_alignment = std::alignment_of<Vec4T>::value;
```

Comment:
We actually don't have any real vectorized kernel except the bias_dropout, and it doesn't check the alignment there, which is a bug. But since currently our tensor is always contiguous, and CUDA always allcoate memeory that are alignment to 256, so it's not a problem for now. This is not the reason for no pef gain for vectorized kernel, if the memory is not aligned, your code will not be executed.